### PR TITLE
WI #2016 Implement unreachable code detection

### LINKED ISSUE #2016: CFG Detect unreachable Code
**Description**
With the Control Flow Graph we can detect unreachable code. This issue shall implement this feature, so that we can add a warning in the client.

image

### PR DESCRIPTION
**Pull request by fm-117**
Fix #2016.

This is a first attempt at unreachable code detection, it should be improved with the support of multiline diagnostics (see #1846) to report unreachable code ranges just like RdZ does.

### CODE REVIEW

File: TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -1327,12 +1332,68 @@ private void HandlePerformsWithTestAfter()
...
+        /// Traverse the CFG to detect unreachable code blocks.
+        /// </summary>
+        private void DetectUnreachableCode()
+        {
+            //DFS to flag all reachable statements, the hashset is needed because we also traverse BasicBlockForNodeGroups
+            HashSet<int> visitedBlocks = new HashSet<int>();
+            this.CurrentProgramCfgBuilder.Cfg.DFS(FlagReachableStatements);
+
+            //Collect unreachable blocks from remaining unreachable statements
+            HashSet<int> unreachableBlocks = new HashSet<int>();
```
Comment (mayanje) [inline TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs:1345]: We can also store the unreachableBlocks information as a result in the related CFG? beacuse here this informatiion is lost, if one does not check each diagnostic.

Comment (fm-117) [reply TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs:1345]: Ok. Done here 1cc51fbf07085027fd5a1f80c40f7099af8c49c6.

**Review by mayanje — APPROVED** [review]

**Review by delevoye — CHANGES_REQUESTED** [review]

File: TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -1327,12 +1332,70 @@ private void HandlePerformsWithTestAfter()
...
+            //TODO We should report unreachable ranges instead, waiting for #1846 to be solved first...
+            foreach (var unreachableBlockIndex in unreachableBlocks)
+            {
+                var unreachableBlock = this.CurrentProgramCfgBuilder.Cfg.AllBlocks[unreachableBlockIndex];
+                this.CurrentProgramCfgBuilder.Cfg.AddUnreachableBlock(unreachableBlock);
+
+                var firstInstruction = unreachableBlock.Instructions.First.Value;
+                var diagnostic = new Diagnostic(MessageCode.Warning, firstInstruction.CodeElement.Position(), "Unreachable code detected");
+                AddDiagnostic(diagnostic);
+            }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs:1366]: Well not exactly, in the previous foreach, we may have duplicate of the same block so we still need the HashSet to keep track of unique unreachable blocks. But you're right, we can merge the two loops, using the HashSet only as a marker and not iterating over it.

File: TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -146,6 +146,8 @@ private Sentence CurrentSentence
         /// </summary>
         private List<Sentence> AllSentences;
 
+        private Dictionary<Node, bool> AllStatements;
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs:153]: Yrs, done in 2280c097e18198128164b77bdb86a287df3c2701.

File: TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs]: Ok, done in 2280c097e18198128164b77bdb86a287df3c2701.

**Review by delevoye — APPROVED** [review]

**Pull request merged**